x/ibc/02-client/keeper/proposal.go

```
@@ -0,0 +1,74 @@
+// In case 2) before trying to update the client, the client will be unfreeze by calling Unfreeze().
+// Note, that even if the update happens, it may not be successful.
+// The reason is that there are still some validation checks on the header that may fail and
+// throw and error before the update is completed.
+func (k Keeper) ClientUpdateProposal(ctx sdk.Context, p *types.ClientUpdateProposal) error {
+	clientState, found := k.GetClientState(ctx, p.ClientId)
+	if !found {
+		return types.ErrClientNotFound
+	}
+
+	// TODO: it will be nice to refactor the following piece of code
+	// and add a function to the clientStatus interface in order to
+	// avoid to use the switch statement. A naive idea is to add
+	// a function with following signature:
+	//
+	// TryClientUpdateProposal(sdk.Context, k keeper.Keeper, *types.ClientUpdateProposal) error
+	//
+	// Unfortunately, we can't pursuit this route as of now, because we face the challenge of an import cycle
+	// with the keeper package which is not easy to resolve.
+	clientType := clientState.ClientType()
+	switch clientType {
+	case exported.Tendermint:
+
+		tmClientState := clientState.(*ibctmtypes.ClientState)
+
+		updateClientFlag := false
+		overrideFlag := false
+
+		if tmClientState.AllowGovernanceOverrideAfterExpiry && tmClientState.Expired(ctx.BlockTime()) {
+			updateClientFlag = true
+			overrideFlag = true
+		}
+
+		if tmClientState.AllowGovernanceOverrideAfterMisbehaviour && tmClientState.IsFrozen() {
+			tmClientState.Unfreeze()
+			k.SetClientState(ctx, p.ClientId, tmClientState)
+			updateClientFlag = true
+		}
+
+		if updateClientFlag {
+
+			var tmtHeader ibctmtypes.Header
+			h, err := tmtHeader.UnmarshalBinaryBare(p.Header)
+			if err != nil {
+				return types.ErrInvalidHeader
+			}
+			if _, err = k.UpdateClient(ctx, p.ClientId, h, overrideFlag); err != nil {
+				return err
+			}
+
+		} else {
+			return types.ErrFailUpdateClient
+		}
```

```suggestion
		if !updateClientFlag {
			return types.ErrFailUpdateClient
		}
		var tmHeader ibctmtypes.Header
		header, err := tmHeader.UnmarshalBinaryBare(p.Header)
		if err != nil {
			return types.ErrInvalidHeader
		}
		
		if _, err = k.UpdateClient(ctx, p.ClientId, header, overrideFlag); err != nil {
			return err
		}
```

---

although I don't understand why do we need to return an error if `updateClientFlag` is false